### PR TITLE
bpo-38870: Implement Simple Preceding to AST Unparser

### PR DESCRIPTION
**Pull request by isidentical**
cc: @pablogsal 

https://bugs.python.org/issue38870

### CODE REVIEW

**Comment by isidentical** [comment]
Do we need news entry for this fix?

**Comment by isidentical** [comment]
I am not sure why this exists and what it does but it seems wrong to me, at least with checking everything in ast. 
https://github.com/python/cpython/blob/ded8888fbc33011dd39b7b1c86a5adfacc4943f3/Lib/test/test_ast.py#L326-L331
It was failing because of both `_Levels` and `IntEnum`. I first thinked an extra condition about if name not starts with underscore but it doesnt worked because of `IntEnum`. So I moved instance of AST check upper.

File: Lib/ast.py
Comment (eric-wieser) [inline Lib/ast.py]: Perhaps clearer to drop `_Delimit.condition` here, and use instead:
```
def delimit_if(self, condition, delimiter):
    if condition:
        return self._Delimit(self, delimiter)
    else:
        return contextlib.nullcontext()
```

File: Lib/ast.py
Comment (eric-wieser) [inline Lib/ast.py]: Perhaps better to make this a `set`

File: Lib/ast.py
Comment (isidentical) [reply Lib/ast.py]: unfortunately we can't use contextlib due to performance issues.

File: Lib/ast.py
Comment (isidentical) [reply Lib/ast.py]: True, thanks.

File: Lib/ast.py
Comment (eric-wieser) [reply Lib/ast.py]: Still tempting to create separate classes, which gives you one branch and not two:
```python
class _NoDelimit:
    def __enter__(self): pass
    def __exit__(self, *args): pass

class _Delimit:
    """A context manager for preparing the source for expressions. It adds
    start of the delimiter  and enters, after exit it adds delimiter end."""
    def __init__(self, unparser, delimiter):
        self.unparser = unparser
        self.delimiter = delimiter
    def __enter__(self):
        self.unparser.write(self.delimiter[0])
    def __exit__(self, exc_type, exc_value, traceback):
        self.unparser.write(self.delimiter[-1])

def delimit_if(self, condition, delimiter):
    if condition:
        return self._Delimit(self, delimiter)
    else:
        return _NoDelimit()
```

Comment (isidentical) [reply Lib/ast.py]: This is going to make this code more complex and I am not sure if we keep going like this, @pablogsal what do you think about re-bringing contextlib back. Then we can just return `nullcontext` and even better we can just use `contextmanager`

**Comment by isidentical** [comment]
@pablogsal is it worth to mention there will be an extra field in certain nodes called `_precedence_level` so after unparsing the tree will be mutated. Maybe we can strip out that field by traversing tree with `ast.walk` after generating result? (but that would definitely slow down)

**Review by pablogsal — CHANGES_REQUESTED** [review]
I still didn't have time to go to the full PR but some general comments:

* I find the diff very difficult to read. You rearranged many functions and change signatures so the diff is bigger than it should be and that makes it complicated to read the changes and how the feature is implemented. Please, try to change only the relevant things for this issue in particular.

* The class should *not* modify the ast object that you pass into or mutate it in any way, so please, do not attach attributes to the nodes (like `_precedence_level`) or do any mutation over the tree. 

* Bringing enum into the `ast` module will likely introduce another performance regression (please, measure).

**Comment by isidentical** [comment]
I have made the requested changes; please review again

**Comment by pablogsal** [comment]
I'm currently handling some urgent regressions on buildbots, so it will take me some time to come back to this :(

File: Lib/ast.py
Comment (vstinner) [inline Lib/ast.py]: You may use an enum for this.

File: Lib/ast.py
Comment (vstinner) [inline Lib/ast.py]: To make the PR more readable, I suggest to write first PR which only added delim() context manager but a version without argument. A PR basically only to change the indentation. Once it's merged, rebased this PR on top of it.

A PR just to add "(" an ")" as context manager sounds like a good idea to me, even if this PR is not accepted.

**Comment by isidentical** [comment]
Thanks for reviews @vstinner, can I get an `DO NOT MERGE` label until the other PR I'm going to open is merged? It is going to look messy until then.

**Comment by vstinner** [comment]
> Thanks for reviews @vstinner, can I get an DO NOT MERGE label until the other PR I'm going to open is merged? It is going to look messy until then.

You can add [WIP] to the PR title.

File: Lib/ast.py
Comment (pablogsal) [reply Lib/ast.py]: Importing `enum` will make importing the module slower, and some people were concerned about that.

Comment (isidentical) [reply Lib/ast.py]: Is there any chance we can split `ast` with `unparse` like `collections` and `collections.abc`? I believe in that way code will be much more readable and easy to maintain.

Comment (pablogsal) [reply Lib/ast.py]: I don't think having a module for only one function is worth enough. At that point the trade-off of the imports becomes similar

Comment (vstinner) [reply Lib/ast.py]: Well, that's why I proposed to put the unparse feature in a separated ast.unparse submodule in the first place ;-)

Comment (pablogsal) [reply Lib/ast.py]: > Well, that's why I proposed to put the unparse feature in a separated ast.unparse submodule in the first place ;-)

I still think is not a good idea to have a separate module for just one function being the reason just that we want to import things.  From `ast.unparse import unparse` feels weird if there is nothing else there (and I still think this is not a good justification to also expose the underlying class).

Comment (isidentical) [reply Lib/ast.py]: @pablogsal if `ast.unparse import unparse` feels weird, what about lazy loading. We can create a module named `_ast_unparse` etc. and then load `unparse` function if needed by module level getattr.

**Comment by isidentical** [comment]
@pablogsal @vstinner #17612 is merged and I rewrote this preceding patch with basing that delimiting changes. The diff is much more minimal right now. This patch tries to implement the preceding system that was also used in `ast_unparse.c` and aims to remove unnecesary parens arround the expressions.

File: Lib/ast.py
Comment (pablogsal) [inline Lib/ast.py]: Should this method be a classmethod? ` type(self)()` looks suspicious especially since we don't use `self` in the function anymore.

File: Lib/test/test_ast.py
Comment (pablogsal) [inline Lib/test/test_ast.py]: This guard is already too big, I suggest we factor into a function that describes what we are actually checking for?

File: Lib/test/test_unparse.py
Comment (pablogsal) [inline Lib/test/test_unparse.py:313]: Let's add some arithmetic operations with parenthesis (like `(1+2)*3+4*(5+2)`).

File: Lib/ast.py
Comment (pablogsal) [inline Lib/ast.py]: Nit: Align the comment

Comment (isidentical) [reply Lib/ast.py]: ups, thanks

File: Lib/test/test_unparse.py
Comment (isidentical) [reply Lib/test/test_unparse.py:313]: sure

File: Lib/test/test_ast.py
Comment (isidentical) [reply Lib/test/test_ast.py]: I dont think I can describe what it does but I'll refactor it into a guard function and use it.

Comment (pablogsal) [reply Lib/test/test_ast.py]: Something like  `is_valid_ast_test_class` is enough.

File: Lib/test/test_ast.py
Comment (pablogsal) [inline Lib/test/test_ast.py]: In this way, this is a bit deceiving because both `ifs` could be false and then the function returns `None`. Even if this cannot happen I find it suboptimal. 

Also, this second test is not what it was before:

```
        if node is not ast.AST and issubclass(node, ast.AST):
```

I would suggest to just move the check as it is into a function as it is for now.

Comment (isidentical) [reply Lib/test/test_ast.py]: What is wrong with returning `None`, it is still falsy. But for clarifying, I can add a return False.

Comment (isidentical) [reply Lib/test/test_ast.py]: Oh my bad. I still dont understand what is the purpose of that test so I might did refactoring wrong, sorry. I'll try to do it again

File: Lib/ast.py
Comment (pablogsal) [inline Lib/ast.py]: You never use the `by` argument, so why have it?

File: Lib/test/test_ast.py
Comment (pablogsal) [inline Lib/test/test_ast.py]: You don't need here the `isinstance(node, type)` because if is false it would have exited in the first branch, no?

Comment (isidentical) [reply Lib/test/test_ast.py]: right, I forgot to remove it.

File: Lib/ast.py
Comment (isidentical) [reply Lib/ast.py]: I was using it in here 
https://github.com/isidentical/cpython/blob/4576f36413ff5dd53b54674c68212cc3f95ff2be/Lib/ast.py#L1143-L1148
like this
```py
            left_precedence = operator_precedence.next(by=int(rassoc))
            right_precedence = operator_precedence.next(by=int(not rassoc))
```
but I believe current implementation is simpler. If you agree, I can remove that parameter and keep the current one

Comment (pablogsal) [reply Lib/ast.py]: > but I believe current implementation is simpler. If you agree, I can remove that parameter and keep the current one

I would suggest to remove it as is not clear to me what it does without checking the function (maybe a better name will do, but I think a function call `next` should not do much more than just give me the next value :)

File: Lib/ast.py
Comment (vstinner) [inline Lib/ast.py]: That sounds overkill, why not just testing if it's equal to ATOM?

Comment (eric-wieser) [reply Lib/ast.py]: Or perhaps use `try/except` to catch the `ValueError` that `type(self)(self + 1)` raises, and return `self`

Comment (isidentical) [reply Lib/ast.py]: Thanks @vstinner, @eric-wieser. I was thinking this might be changed through the time but I agree it looks like an overkill.

**Review by pablogsal — APPROVED** [review]
This looks good to me with the latest corrections!

Great job, @isidentical! :tada:

**Comment by vstinner** [comment]
Well done @isidentical. I'm happy to see that the merged change now uses the enum module :-)

**Pull request merged**
